include/GoogleSync/GoogleSyncBase.php

```diff
@@ -0,0 +1,1049 @@
+ * of this program must display Appropriate Legal Notices, as required under
+ * Section 5 of the GNU Affero General Public License version 3.
+ *
+ * In accordance with Section 7(b) of the GNU Affero General Public License version 3,
+ * these Appropriate Legal Notices must retain the display of the "Powered by
+ * SugarCRM" logo and "Supercharged by SuiteCRM" logo. If the display of the logos is not
+ * reasonably feasible for technical reasons, the Appropriate Legal Notices must
+ * display the words "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ */
+if (!defined('sugarEntry') || !sugarEntry) {
+    die('Not A Valid Entry Point');
+}
+
+require_once __DIR__ . '/../../modules/Users/User.php';
+require_once __DIR__ . '/../../modules/Meetings/Meeting.php';
+require_once __DIR__ . '/GoogleSyncExceptions.php';
+
+use SuiteCRM\Utility\SuiteValidator;
+
+/**
+ * Implements Google Calendar Syncing
+ *
+ * @license https://raw.githubusercontent.com/salesagility/SuiteCRM/master/LICENSE.txt
+ * GNU Affero General Public License version 3
+ * @author Benjamin Long <ben@offsite.guru>
+ */
+
+class GoogleSyncBase
+{    
+    /** @var User The SuiteCRM User Bean we're currently working with */
+    protected $workingUser;
+
+    /** @var \Google_Client The Google client object for the current sync job */
+    protected $gClient;
+
+    /** @var \Google_Service_Calendar The Google Calendar Service Object */
+    protected $gService;
+
+    /** @var array The Google AuthcConfig json */
+    protected $authJson = array();
+
+    /** @var string The local timezone */
+    protected $timezone;
+
+    /** @var string The Calendar ID */
+    protected $calendarId;
+
+    /** @var array An array of SuiteCRM meeting id's that we've already synced this session */
+    protected $syncedList = array();
+
+    /** @var object A Database Instance */
+    protected $db;
+
+    /** @var object A Logger Instance */
+    protected $logger;
+
+    /** @var string The log level before we begin */
+    protected $oldLogLevel;
+
+    /**
+     * Class Constructor
+     * 
+     * @param array $sugarConfig - using $sugar_config as a dependency
+     */
+    public function __construct($sugarConfig)
+    {
+        // This sets the log level to a variable that can be set on the command line while running cron.php on the server. It's for debugging only.
+        // EXAMPLE: $ GSYNC_LOGLEVEL=debug php cron.php
+        $this->logger = LoggerManager::getLogger();
+        if (isset($_SERVER['GSYNC_LOGLEVEL'])) { // TODO: is it possible to using $sugar_config instead superglobals? (a global usage is one step closer to clean code than using a superglobal)
+            $this->oldLogLevel = $this->logger->getLogLevel();
+            $this->logger->setLevel($_SERVER['GSYNC_LOGLEVEL']); // TODO: it changes the log level but as it is a global, it will affect the other parts of core until this class __destructor is called! (and it's bad)
```

Comment:
@gymad 
Yes. That is by design. It's only used for debugging. With this I can have a user set the debug level only for the sync process. It will log at that level for all child functions. This way they only see the debug log, for instance, of the sync job.
I don't think that this sets the log level globally for the whole system. It will only set it for children of this function. Am I incorrect? I've never seen anything other than sync related logging when using this. Nothing else.

Reply:
Two point is not clear for me here:
If it's needed for debugging only why not just simply check for e.g the isDeveloperMode() function instead introducing a new superglobal. Or for e.g you also can check a $sugar_config variable which is specially for this and add it to your config_override.php instead introduce new superglobals...
Second problem what I can see is you restore the log level only in destructor. It means everything (but really everything not just your new code and functions) will using the log level as defined in `$_SERVER['GSYNC_LOGLEVEL']` after the constructor called until this destructor set it back. 
Maybe more clear if I take this into an example: 

Here is a very simple php file (only for example..)
```
<?php
hereDoingThings();  // <-- using original log level. .. OK :)
?>
```

Now I want to add your code into this because at this php I want to Sync my google accouns but I don't want to brake anything in original behaviour of this but I will do. (In other words, where your destructor called in the following example php file after this modification?)

```
<?php
 // it will change the log level:
$sync = new GoogleSync(..);
$sync->doingThingsWithNewLogLevel(...); // using debug level .. OK

// here everything else using your specified log level but shouldn't.
// It because the LoggerManager::$_instance is static pointer to a logger instance
// and LoggerManager::getLogger() is a singleton factory (antipattern unfortunatelly)

hereDoingThings();  // <-- using debug level, however I did not change it .. BAD :(

$sync->doingSomeOtherThings(); // still debug level.. OK

// php file ended, GoogleSyncBase::__destruct called and restoring log level only at this point.
?>
```

As you see, the logger badly implemented as a static singleton. I know it is a wrong desing but unfortunatelly it's already added in this way and it means we should be carefull.

Reply:
@gymad Ok. I understand now. The Logger is not an independent object, as I thought it was. The whole LoggerManager class is forcing this. Ugh...
Even __construct is a private method. :man_facepalming:

How about if I explicitly destroy the class after running the sync by adding '$googleSync = null' in the syncGoogleCalendar() function modules/Schedulers/_AddJobsHere.php ?
Would that be enough to keep this logging functionality? I don't know any other way around this. I'll have to remove the custom logging.

PS. I had already changed from using a superglobal to an entry in config_override.php.

Reply:
I've removed the logging code.